packages/astro/src/vite-plugins/override-components.ts

```diff
@@ -0,0 +1,72 @@
+/**
+ * A plugin that lets users to override TutorialKit's components.
+ *
+ * The virtual module can be imported as:
+ *
+ * ```ts
+ * import { TopBar } from 'tutorialkit:override-components';
+ *
+ * <TopBar />
+ * ```
+ *
+ * User can override the components in `astro.config.ts`:
+ *
+ * ```ts
+ * export default defineConfig({
+ *   integrations: [
+ *     tutorialkit({
+ *       components: {
+ *         TopBar: './CustomTopBar.astro',
+ *       },
+ *     }),
+ *   ],
+ * });
+ * ```
+ */
+import type { VitePlugin } from '../types.js';
+
+export interface OverrideComponentsOptions {
+  /**
+   * Component for overriding the top bar.
+   *
+   * This component has 3 slots that are used to pass TutorialKit's default components:
+   * - `logo`: Logo of the application
+   * - `theme-switch`:  Switch for changing the theme
+   * - `login-button`: TODO, add description here
+   *
+   * Usage:
+   * ```jsx
+   *   <slot name="logo" />
+   *   <slot name="theme-switch" />
+   *   <slot name="login-button" />
+   * ```
+   */
+  TopBar?: string;
+}
+
+const virtualModuleId = 'tutorialkit:override-components';
+const resolvedId = `\0${virtualModuleId}`;
+
+export function overrideComponents(options?: OverrideComponentsOptions): VitePlugin {
+  const topBar = options?.TopBar || '@tutorialkit/astro/default/components/TopBar.astro';
```

Comment:
This should also check whether `defaultRoutes` is false or not. If it's `false`, then we don't want to use the one from `@tutorialkit/astro` but the one from `src/components/TopBar.astro`.

This is needed to make sure that `eject` works as expected.

Note that I would agree that they should probably remove the import to `tutorialkit:override-components`, after doing `eject`. If you prefer you could also modify `eject` to do so.

Reply:
I think it makes more sense to modify the import path in source files, rather than the plugin itself. I'll check how to do this during `eject` command.

Reply:
I've added support for `tutorialkit eject` command now. During that we'll:
- Remove `components: {...}` from `astro.config.ts`
- Replace `tutorialkit:override-components` imports with relative file imports. If user had defined custom `TopBar` in `astro.config.ts` we'll use that one. If not, use default one. 

I'll be updating and adding test cases next.

Reply:
> Can you elaborate why? The plugin is internal and the change is really simple, no?

To me it makes sense that imports from virtual module `tutorialkit:override-components` are removed during `tutorialkit eject`. If they are left as is, we would consider the internal Vite plugin as part of the public API. What do you think?

> I prefer the copy

We do still copy all the files as is. Now we just replace the import from virtual module to the actual file.

Reply:
> To me it makes sense that imports from virtual module `tutorialkit:override-components` are removed during `tutorialkit eject`. If they are left as is, we would consider the internal Vite plugin as part of the public API. What do you think?

Ah good point, I didn't thought about that.

It might not be too bad though given how simple the plugin is and I don't see it changing much. That plus the fact that the Astro integration cannot be removed after doing an `eject` as all the other plugins it adds are needed.

However with having the renaming on disk, now the name of the file becomes part of the public API which in my opinion is much harder to maintain.

> Now we just replace the import from virtual module to the actual file.

I think this is where it sounds great on paper, but long term it won't work well because:

 * Right now the CLI looks at the `@tutorialkit/astro/dist/default` folder from their project and just copy the files. We do this because we don't want to inject incompatible code and also because we don't _care_ about it's setup, we now it works when `defaultRoutes: true`, it will work when `defaultRoutes: false`.
 * The Astro integration might be a different version of the CLI, so maybe the CLI will forget to rename some components because it's too old, maybe it will crash thinking that a rename would work when it doesn't because the project is too old.
 * What if in the future we rename those files? Then the CLI will need to maintain all names in mind, and again if they use a CLI that's too old (which if they run `npx tutorialkit eject` would likely be the case) then it wouldn't work.

The simple copy is forward / backward compatible because it doesn't care about the internals. It just move files around in their project and makes no assumption about them.

Reply:
> What if in the future we rename those files? Then the CLI will need to maintain all names in mind, and again if they use a CLI that's too old

We could resolve the name from `@tutorialkit/astro/default/components/TopBar.astro` entrypoint. Then it would remain up-to-date always. Removing that requires major version release as it's part of package's `exports`.

If we don't remove `tutorialkit:override-components` from source code that's generated to file system during `tutorialkit eject`, what's the alternative? Add the same resolving logic to the plugin? Something like below?

```diff
export function overrideComponents(options?: Options): VitePlugin {
+  // Fallback to local relative file when `defaultRoutes` is disabled, otherwise use public entrypoint
+  const fallback = options?.defaultRoutes === false ? './TopBar.astro' : '@tutorialkit/astro/default/components/TopBar.astro';
+
+  const topBar = options?.components?.TopBar || fallback;
-  const topBar = options?.components?.TopBar || '@tutorialkit/astro/default/components/TopBar.astro';
```

Reply:
> We could resolve the name from `@tutorialkit/astro/default/components/TopBar.astro` entrypoint. Then it would remain up-to-date always. Removing that requires major version release as it's part of package's `exports`.

That would help a bit but it wouldn't solve the CLI / project version mistmatch. You would still need to use the same version between the two or it would break.

I suppose we could also throw an error if the version of `tutorialkit` in the project does not match the version of the CLI or install the correct CLI and `execve` into that.

> If we don't remove `tutorialkit:override-components` from source code that's generated to file system during `tutorialkit eject`, what's the alternative? Add the same resolving logic to the plugin? Something like below?

Yes, this is what I had in mind (or if it doesn't work use the `config.projectRoot` to resolve the path to an absolute path)